lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,288 @@
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull()
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            //could have ID feed into [ID] (as a possible batch hydration case).
+            // in this case we need to unwrap the list and check types
+            if (isBatchHydration && (!unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList)) {
+                val error = getHydrationInputErrors(
+                        unwrappedHydrationSourceFieldType,
+                        unwrappedActorFieldArgType.unwrapOne(),
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+                if (error != null) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+            }
+            //could have [ID] feed into ID (non-batch, ManyToOne case explained in NadelHydrationStrategy.kt)
+            // in this case we need to unwrap the list and check types
+            else if (!isBatchHydration && (unwrappedHydrationSourceFieldType.isList && !unwrappedActorFieldArgType.isList)) {
+                val error = getHydrationInputErrors(
+                        unwrappedHydrationSourceFieldType.unwrapOne(),
+                        unwrappedActorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+                if (error != null) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+
+            }
+            // Otherwise we can just check the types normally
+            else {
+                return getHydrationInputErrors(
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+            }
+            return null
```

Comment:
This should be baked into the if statements above, it makes it a lot easier to follow.

It should read like

```kotlin
return if () {
  getError()
} else if () {
  getError()
} else {
  getError()
}
```

But right now it's


```kotlin
if () {
  val error = getError()
  if (error != null) {
    return error
  }
} else if () {
  val error = getError()
  if (error != null) {
    return error
  }
} else {
  val error = getError()
  return error
}

return null
```